io/io/src/TFile.cxx

```diff
@@ -1474,27 +1474,55 @@ Bool_t TFile::IsOpen() const
 
 void TFile::MakeFree(Long64_t first, Long64_t last)
 {
-   TFree *f1      = (TFree*)fFree->First();
-   if (!f1) return;
-   TFree *newfree = f1->AddFree(fFree,first,last);
-   if(!newfree) return;
+   TFree *f1 = static_cast<TFree *>(fFree->First());
+   if (!f1)
+      return;
+
+   TFree *newfree = f1->AddFree(fFree, first, last);
+   if (!newfree)
+      return;
    Long64_t nfirst = newfree->GetFirst();
    Long64_t nlast  = newfree->GetLast();
-   Long64_t nbytesl= nlast-nfirst+1;
-   if (nbytesl > 2000000000) nbytesl = 2000000000;
-   Int_t nbytes    = -Int_t (nbytesl);
+
+   // The new free segment may close a series of consecutive free segments at the end of the file.
+   // These segments need to be merged so that the last free segment is [fEND + 1 .. max file size].
+   auto tail = static_cast<TFree *>(fFree->Last());
+   while (tail != fFree->First()) {
+      // Starting from the last segment, merge in previous segments if they are adjacent
+      auto prev = (TFree *)fFree->Before(tail);
+      if (prev->GetLast() + 1 < tail->GetFirst())
+         break;
+      assert(prev->GetLast() + 1 == tail->GetFirst());
+      tail->SetFirst(prev->GetFirst());
+      fFree->Remove(prev);
+      delete prev;
+   }
+   if (tail->GetFirst() <= nfirst) {
+      nfirst = tail->GetFirst();
+      nlast = tail->GetLast();
+   }
+
+   Long64_t nbytesl = nlast - nfirst + 1;
+   if (nbytesl > TFree::kMaxGapSize)
+      nbytesl = TFree::kMaxGapSize;
```

Comment:
If I understand correctly, here we are "sacrificing" `nbytesl - kMaxGapSize` bytes and leaving them unindexed; shouldn't we also add a new free segment right after `tail` to encompass them if this is not at the end of the file?

Reply:
(side note, this lost was already present is the current code).